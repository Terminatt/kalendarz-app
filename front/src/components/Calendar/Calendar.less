@import '~styles/variables.less';

.calendar {
  width: 800px;
  background-color: @primary-bg;
  padding: @padding-small;
  border-radius: @border-radius-regular;
  box-shadow: @box-shadow-mild;

  &-dates {
    display: grid;
    grid-template-columns: repeat(7, 1fr);
    grid-auto-rows: 80px;
    gap: 26px;
  }
}