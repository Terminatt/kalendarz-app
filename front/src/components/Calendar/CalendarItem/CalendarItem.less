@import '~styles/variables.less';

.calendar-item {
  background-color: @secondary-bg;
  border: none;
  border-radius: @border-radius-calendar;
  box-shadow: @box-shadow-tiny;
  border-left: 3px solid @border-color-base;
  cursor: pointer;
  
  &-normal {
    border-color: @primary-color;
  }

  &-another-month-day {
    border-color: @calendar-day-disabled;
    color: @calendar-day-disabled;
    cursor: not-allowed;
  }

  &-before-today {
    border-color: @calendar-day-disabled;
  }

  &-today {
    border-color: @link-color;
  }

  &-dayoff {
    border-color: @red-color;
  }
}