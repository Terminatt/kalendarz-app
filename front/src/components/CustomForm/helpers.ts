import { ResponseError } from '@generics/generics';
import { getEntries } from '@utils/general';

export function parseErrorResponse(errorResponse: ResponseError, ERROR_MESSAGES: {[key: string]: string}):
Array<{name: string, errors: string[]}> {
    const entries = getEntries(errorResponse);
    const errors: Array<{name: string, errors: string[]}> = [];

    // TODO SERVE ARRAY CASE
    entries.forEach((el) => {
        if (el[1] && !Array.isArray(el[1])) {
            errors.push({ name: el[0], errors: [ERROR_MESSAGES[el[1].type]] });
        }
    });

    return errors;
}
