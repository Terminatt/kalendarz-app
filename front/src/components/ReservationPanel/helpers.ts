import { TIME_BLOCK_MINUTES } from '@constants/constants';
import { HoveredFocusedTimeBlock, TimeInterval } from './ReservationPanel';

export function getTimeBlockValue(timeBlockIndex: number): string {
    const timeValue = TIME_BLOCK_MINUTES * timeBlockIndex;
    return timeValue === 0 ? '00' : timeValue.toString();
}

export function isTimeBlockSelected(timeValue: number, interval?: TimeInterval, hoveredFocused?: HoveredFocusedTimeBlock): boolean {
    if (!interval?.start) {
        return false;
    }

    if (hoveredFocused?.hovered) {
        return timeValue <= hoveredFocused.hovered;
    }

    if (hoveredFocused?.focused) {
        return timeValue <= hoveredFocused.focused;
    }

    if (interval?.end) {
        return timeValue >= interval.start && timeValue <= interval.end;
    }

    return timeValue === interval.start;
}
