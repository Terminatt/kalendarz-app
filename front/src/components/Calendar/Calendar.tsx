import { isExisting, joinClassNames } from '@utils/general';
import dayjs, { Dayjs } from 'dayjs';
import React, { useCallback, useEffect, useState } from 'react';
import { Month, monthNames } from '@constants/constants';
import Switcher, { IndexExceed } from '@components/Switcher/Switcher';
import HugeDivider from '@components/HugeDivider/HugeDivider';
import CalendarItem, { CalendarItemType } from './CalendarItem/CalendarItem';
import {
    CalendarDay,
    createDayList,
    createLastWeekList,
    evaluateCurrentMonthDayType,
    getMonthsOptions,
} from './helpers';

import './Calendar.less';

export interface CalendarProps {
    className?: string;
}

const Calendar: React.FC<CalendarProps> = (props) => {
    const [today, setToday] = useState<Dayjs | null>(null);
    const [selectedYear, setYear] = useState<number | null>(null);
    const [selectedMonth, setMonth] = useState<number | null>(null);
    const [daysInMonth, setDaysInMonth] = useState<CalendarDay[]>([]);
    const monthOptions = getMonthsOptions(monthNames);
    const { className } = props;

    const createDaysInMonth = useCallback((_selectedYear: number, _selectedMonth: number): CalendarDay[] => {
        const dayList = createDayList(_selectedYear, _selectedMonth, evaluateCurrentMonthDayType);
        const weekMonthBefore = createLastWeekList(_selectedYear, _selectedMonth - 1, () => CalendarItemType.ANOTHER_MONTH_DAY);

        return [...weekMonthBefore, ...dayList];
    }, []);

    useEffect(() => {
        const currentDate = dayjs();
        setToday(today);
        setYear(currentDate.year());
        setMonth(currentDate.month());
    }, []);

    useEffect(() => {
        if (!isExisting(selectedYear) || !isExisting(selectedMonth)) {
            return;
        }

        setDaysInMonth(createDaysInMonth(selectedYear, selectedMonth));
    }, [selectedMonth]);

    const changeYearOnExceeding = useCallback((exceeds: IndexExceed): void => {
        if (!selectedYear) {
            return;
        }

        if (exceeds === IndexExceed.LEFT) {
            setYear(selectedYear - 1);
            setMonth(Month.DECEMBER);
            return;
        }

        setYear(selectedYear + 1);
        setMonth(Month.JANUARY);
    }, [selectedYear]);

    const onSwitcherChange = useCallback((value: number | null, exceeds?: IndexExceed): void => {
        if (exceeds) {
            changeYearOnExceeding(exceeds);
            return;
        }

        setMonth(value);
    }, [selectedYear]);

    return (
        <div className={joinClassNames(['calendar', className])}>
            <HugeDivider className="calendar-divider" text={selectedYear || ''} />
            <div className="calendar-switcher">
                <Switcher selected={selectedMonth} onChange={onSwitcherChange} options={monthOptions} />
            </div>
            <div className="calendar-content">
                <div className="calendar-content-header">
                    <h2>Wybierz dzie≈Ñ</h2>
                </div>
                <div className="calendar-content-dates">
                    {daysInMonth.map((el) => (
                        <CalendarItem key={el.dayNumber + el.type} type={el.type} dayName={el.dayName} dayNumber={el.dayNumber} />
                    ))}
                </div>
            </div>
        </div>
    );
};

export default Calendar;
