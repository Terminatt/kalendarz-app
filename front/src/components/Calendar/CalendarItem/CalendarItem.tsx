import React, { useCallback } from 'react';

import './CalendarItem.less';

export enum CalendarItemType {
  ANOTHER_MONTH_DAY = 'another-month-day',
  BEFORE_TODAY = 'before-today',
  TODAY = 'today',
  NORMAL = 'normal',
  DAYOFF = 'dayoff',
}

export interface CalendarItemProps {
  type: CalendarItemType,
  dayNumber: string;
  dayName: string;
  onClick?: () => void;
}

const CalendarItem: React.FC<CalendarItemProps> = (props) => {
    const {
        type, dayNumber, dayName, onClick,
    } = props;
    const disabled = type === CalendarItemType.ANOTHER_MONTH_DAY || type === CalendarItemType.DAYOFF;

    const onItemClick = useCallback(() => {
        if (!onClick) {
            return;
        }
        onClick();
    }, [onClick]);

    return (
        <button onClick={onItemClick} className={`calendar-item calendar-item-${type}`} type="button" disabled={disabled}>
            <div>{dayNumber}</div>
            <div>{dayName}</div>
        </button>
    );
};

export default CalendarItem;
