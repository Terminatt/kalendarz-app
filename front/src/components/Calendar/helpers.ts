import { SwitcherOption } from '@components/Switcher/Switcher';
import { DAY_NAMES, Month } from '@constants/constants';
import {
    isBeforeToday, isToday, isWeekend,
} from '@utils/general';
import dayjs, { Dayjs } from 'dayjs';
import { CalendarItemType } from './CalendarItem/CalendarItem';

export interface CalendarDay {
    dayNumber: string;
    dayName: string;
    type: CalendarItemType;
    date: Dayjs;
}
export type EvaluateTypeHandler = (date: Dayjs) => CalendarItemType;
export function createDayList(year: number, month: Month, evaluateType: EvaluateTypeHandler): CalendarDay[] {
    const dayList: CalendarDay[] = [];
    const date = new Date(year, month, 1);
    const numberOfDays = dayjs(date).daysInMonth();

    for (let i = 0; i < numberOfDays; i++) {
        const dateTmp = dayjs(date).add(i, 'day');
        const dayNumber = dateTmp.day();
        dayList.push(
            {
                dayNumber: i < 9 ? `0${i + 1}` : `${i + 1}`,
                dayName: DAY_NAMES[dayNumber],
                type: evaluateType(dateTmp),
                date: dateTmp,
            },
        );
    }
    return dayList;
}

export function createLastWeekList(year: number, month: Month, evaluateType: EvaluateTypeHandler): CalendarDay[] {
    const dayList: CalendarDay[] = [];
    const date = new Date(year, month, 1);
    const lastDay = dayjs(date).endOf('month');
    const lastDayNumber = lastDay.day();

    for (let i = lastDayNumber - 1; i >= 0; i--) {
        const dateTmp = dayjs(lastDay).subtract(i, 'day');
        const dayNumber = dateTmp.day();
        dayList.push(
            {
                dayNumber: `${dateTmp.format('DD')}`,
                dayName: DAY_NAMES[dayNumber],
                type: evaluateType(dateTmp),
                date: dateTmp,
            },
        );
    }
    return dayList;
}

export function evaluateCurrentMonthDayType(current: Dayjs): CalendarItemType {
    if (isBeforeToday(current)) {
        return CalendarItemType.BEFORE_TODAY;
    }

    if (isToday(current)) {
        return CalendarItemType.TODAY;
    }

    if (isWeekend(current)) {
        return CalendarItemType.DAYOFF;
    }

    return CalendarItemType.NORMAL;
}

export function getMonthsOptions(months: string[]): SwitcherOption<Month>[] {
    return months.map((el, index) => ({ label: el, value: index }));
}
