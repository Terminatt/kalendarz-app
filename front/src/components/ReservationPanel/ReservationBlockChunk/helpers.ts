import { Dayjs } from 'dayjs';
import { TimeIntervalWithRoom } from '../ReservationPanel';

export function isBlockSelected(current: Dayjs, selectedInterval?: TimeIntervalWithRoom, hovered?: Dayjs): boolean {
    if (current.isSame(hovered)) {
        return true;
    }

    if (selectedInterval?.start?.isBefore(current) && selectedInterval?.end?.isAfter(current)) {
        return true;
    }

    if (selectedInterval?.start?.isSame(current) || selectedInterval?.end?.isSame(current)) {
        return true;
    }

    if (!selectedInterval?.start || !hovered) {
        return false;
    }

    if (selectedInterval.startLimit.isAfter(hovered) || selectedInterval.endLimit.isBefore(hovered)) {
        return false;
    }

    if (current.isSameOrAfter(hovered) && current.isBefore(selectedInterval.start)) {
        return true;
    }

    if (!selectedInterval.end) {
        if (current.isSameOrBefore(hovered) && current.isAfter(selectedInterval.start)) {
            return true;
        }
    } else if (current.isSameOrBefore(hovered) && current.isAfter(selectedInterval.end)) {
        return true;
    }

    return false;
}
