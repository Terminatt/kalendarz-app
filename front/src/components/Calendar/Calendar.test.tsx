import React from 'react';
import MockDate from 'mockdate';
import renderer from 'react-test-renderer';
import Calendar from './Calendar';
import { createDayList, createLastWeekList, evaluateCurrentMonthDayType, getMonthsOptions } from './helpers';
import dayjs from 'dayjs';
import { CalendarItemType } from './CalendarItem/CalendarItem';
import { Month, MONTH_NAMES } from '@constants/constants';
import { fireEvent, render } from '@testing-library/react';


const date = '2022-02-20';
describe('Calendar Component', () => {
    
    beforeEach(() => {
        MockDate.set(date);
    });
    
    it('matches the snapshot', () => {
        const tree = renderer
            .create(
                <Calendar />
            )
            .toJSON();
        
            expect(tree).toMatchSnapshot();
    });

    it('triggers click', () => {
        const onDayClick = jest.fn();

        const screen = render(
            <Calendar onDayClick={onDayClick} />
        );
        fireEvent.click(screen.getByText(/01/i));
        expect(onDayClick).toBeCalled();
    })
});

describe('Calendar Component helpers', () => {
    describe('createLastWeekList', () => {
        it('returns days till monday of a month before ', () => {
            const january = createLastWeekList(2022, Month.JANUARY, () => CalendarItemType.ANOTHER_MONTH_DAY);
            expect(january.length).toBe(1);

            const february = createLastWeekList(2022, Month.FEBRUARY, () => CalendarItemType.ANOTHER_MONTH_DAY);
            expect(february.length).toBe(1);

            const march = createLastWeekList(2022, Month.MARCH, () => CalendarItemType.ANOTHER_MONTH_DAY);
            expect(march.length).toBe(4);

            const april = createLastWeekList(2022, Month.APRIL, () => CalendarItemType.ANOTHER_MONTH_DAY);
            expect(april.length).toBe(6);

            const may  = createLastWeekList(2022, Month.MAY, () => CalendarItemType.ANOTHER_MONTH_DAY);
            expect(may.length).toBe(2);

            const june  = createLastWeekList(2022, Month.JUNE, () => CalendarItemType.ANOTHER_MONTH_DAY);
            expect(june.length).toBe(4);

            const july  = createLastWeekList(2022, Month.JULY, () => CalendarItemType.ANOTHER_MONTH_DAY);
            expect(july.length).toBe(0);

            const august  = createLastWeekList(2022, Month.AUGUST, () => CalendarItemType.ANOTHER_MONTH_DAY);
            expect(august.length).toBe(3);

            const september  = createLastWeekList(2022, Month.SEPTEMBER, () => CalendarItemType.ANOTHER_MONTH_DAY);
            expect(september.length).toBe(5);

            const october = createLastWeekList(2022, Month.OCTOBER, () => CalendarItemType.ANOTHER_MONTH_DAY);
            expect(october.length).toBe(1);

            const november = createLastWeekList(2022, Month.NOVEMBER, () => CalendarItemType.ANOTHER_MONTH_DAY);
            expect(november.length).toBe(3);

            const december = createLastWeekList(2022, Month.DECEMBER, () => CalendarItemType.ANOTHER_MONTH_DAY);
            expect(december.length).toBe(6);

        });
    });

    it('getMonthsOptions', () => {
        const options = getMonthsOptions(MONTH_NAMES);

        expect(options.length).toBe(12);
    });

    describe('evaluateCurrentMonthDayType', () => {
        it('returns normal day', () => {
            const type = evaluateCurrentMonthDayType(dayjs(date).add(1, 'day'));

            expect(type).toBe(CalendarItemType.NORMAL);
        });

        it('returns today', () => {
            const type = evaluateCurrentMonthDayType(dayjs(date));

            expect(type).toBe(CalendarItemType.TODAY);
        });

        it('returns another month day', () => {
            const type = evaluateCurrentMonthDayType(dayjs(date).add(-1, 'day'));

            expect(type).toBe(CalendarItemType.BEFORE_TODAY);
        });

        it('returns dayoff', () => {
            const type = evaluateCurrentMonthDayType(dayjs(date).day(7));

            expect(type).toBe(CalendarItemType.DAYOFF);
        });
    });

    describe('createDayList', () => {
        it('returns all days in month', () => {
            const january = createDayList(2022, Month.JANUARY, () => CalendarItemType.NORMAL);
            expect(january.length).toBe(31);

            const february = createDayList(2022, Month.FEBRUARY, () => CalendarItemType.NORMAL);
            expect(february.length).toBe(28);

            // leap year
            const leapFebruary = createDayList(2020, Month.FEBRUARY, () => CalendarItemType.NORMAL);
            expect(leapFebruary.length).toBe(29);

            const march = createDayList(2022, Month.MARCH, () => CalendarItemType.NORMAL);
            expect(march.length).toBe(31);

            const april = createDayList(2022, Month.APRIL, () => CalendarItemType.NORMAL);
            expect(april.length).toBe(30);

            const may  = createDayList(2022, Month.MAY, () => CalendarItemType.NORMAL);
            expect(may.length).toBe(31);

            const june  = createDayList(2022, Month.JUNE, () => CalendarItemType.NORMAL);
            expect(june.length).toBe(30);

            const july  = createDayList(2022, Month.JULY, () => CalendarItemType.NORMAL);
            expect(july.length).toBe(31);

            const august  = createDayList(2022, Month.AUGUST, () => CalendarItemType.NORMAL);
            expect(august.length).toBe(31);

            const september  = createDayList(2022, Month.SEPTEMBER, () => CalendarItemType.NORMAL);
            expect(september.length).toBe(30);

            const october = createDayList(2022, Month.OCTOBER, () => CalendarItemType.NORMAL);
            expect(october.length).toBe(31);

            const november = createDayList(2022, Month.NOVEMBER, () => CalendarItemType.NORMAL);
            expect(november.length).toBe(30);

            const december = createDayList(2022, Month.DECEMBER, () => CalendarItemType.NORMAL);
            expect(december.length).toBe(31);
        })
    })
});
