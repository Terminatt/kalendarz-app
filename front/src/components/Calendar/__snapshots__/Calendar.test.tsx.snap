// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Component matches the snapshot 1`] = `
<div
  className="calendar"
>
  <div
    className="ant-divider ant-divider-horizontal ant-divider-with-text ant-divider-with-text-center huge-divider calendar-divider"
    role="separator"
  >
    <span
      className="ant-divider-inner-text"
    >
      2022
    </span>
  </div>
  <div
    className="calendar-switcher"
  >
    <div
      className="switcher"
    >
      <button
        className="ant-btn transparent-btn switcher-btn switcher-btn-left"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-label="left"
          className="anticon anticon-left"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="left"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
            />
          </svg>
        </span>
      </button>
      <span
        className="switcher-selected"
      >
        Luty
      </span>
      <button
        className="ant-btn transparent-btn switcher-btn switcher-btn-right"
        onClick={[Function]}
        type="button"
      >
        <span
          aria-label="right"
          className="anticon anticon-right"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="right"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
  <div
    className="calendar-content"
  >
    <div
      className="calendar-content-header"
    >
      <h2>
        Wybierz dzie≈Ñ
      </h2>
    </div>
    <div
      className="calendar-content-dates"
    />
  </div>
</div>
`;
